### PR TITLE
Add more error handling to DependencyResolver

### PR DESCRIPTION
**Pull request by ursachec**

### CODE REVIEW

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 1]
```diff
@@ -102,114 +102,113 @@ object GradleDependencies {
...
+          val out                      = new ByteArrayOutputStream()
+          val gradlePropertiesTaskName = "properties"
+          connection
+            .newBuild()
+            .forTasks(gradlePropertiesTaskName)
+            .setStandardOutput(out)
+            .run()
+          out.close()
+          val gradleAndroidPropertyPrefix = "android."
+          !out.toString.split('\n').filter(_.startsWith(gradleAndroidPropertyPrefix)).isEmpty
```
Comment (mpollmeier) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala:126]:
```suggestion
          out.toString.split('\n').filter(_.startsWith(gradleAndroidPropertyPrefix)).nonEmpty
```

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 2]
```diff
@@ -102,114 +102,113 @@ object GradleDependencies {
...
-          project.getTasks.asScala.map(_.getName).toSeq,
-          hasAndroidPrefixGradleProperty
+  ): Try[collection.Seq[String]] = {
+    getGradleProjectInfo(projectDir, projectName) match {
+      case Success(projectInfo) if projectInfo.gradleVersionMajorMinor()._1 < 5 =>
+        Failure(new Exception("Unsupported Gradle version `" + projectInfo.gradleVersion + "`"))
+      case Success(projectInfo) =>
+        logger.info(s"Creating gradle init script...")
+        val destinationDir = Files.createTempDirectory(tempDirPrefix)
+        destinationDir.toFile.deleteOnExit()
```
Comment (mpollmeier) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala:173]: There's a few different ways to create temp files here... I'd suggest using the same everywhere, and would opt for https://github.com/pathikrit/better-files#temporary-files, i.e. 
```
File.usingTemporaryFile { tempFile =>
  //do something
}  // tempFile is auto deleted at the end of this block - even if an exception happens
```

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala (hunk 1 above)
Comment (ursachec) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala]: :+1:

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 3]
```diff
@@ -102,114 +102,113 @@ object GradleDependencies {
...
+      val connection = GradleConnector
+        .newConnector()
+        .forProjectDirectory(projectDir.toFile)
+        .connect()
+      val buildEnv = connection.getModel[BuildEnvironment](classOf[BuildEnvironment])
+      val project  = connection.getModel[GradleProject](classOf[GradleProject])
+
+      val hasAndroidPrefixGradleProperty =
+        try {
+          val out                      = new ByteArrayOutputStream()
```
[the remark this answers is not in the document]
Comment (ursachec) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala]: yes, much better, thx

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 4]
```diff
@@ -114,16 +114,17 @@ object GradleDependencies {
...
-          val gradleAndroidPropertyPrefix = "android."
-          out.toString.split('\n').filter(_.startsWith(gradleAndroidPropertyPrefix)).nonEmpty
+          Using.resource(new ByteArrayOutputStream()) { out =>
+            val gradlePropertiesTaskName = "properties"
+            connection
+              .newBuild()
+              .forTasks(gradlePropertiesTaskName)
+              .setStandardOutput(out)
+              .run()
+            out.close()
```
Comment (max-leuthaeuser) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala:124]: This is done by `Using.resource` itself.

Comment (ursachec) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala]: nice catch, thanks!

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 5]
```diff
@@ -102,114 +102,113 @@ object GradleDependencies {
...
-        case t: Throwable =>
-          logger.warn(s"Caught exception while trying fetch Gradle project information: `$t`.", t)
-          None
-      }
-    if (gradleProjectInfoOption.isEmpty) {
-      throw new Exception("Could not fetch Gradle project information.")
+        Try(GradleConnector.newConnector().forProjectDirectory(projectDir.toFile).connect()) match {
+          case Success(connection) =>
+            val stdoutStream = new ByteArrayOutputStream
+            val stderrStream = new ByteArrayOutputStream
```
[the remark this answers is not in the document]
Comment (ursachec) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala]: :+1:

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 6]
```diff
@@ -102,114 +102,113 @@ object GradleDependencies {
...
+          case t: Throwable =>
+            logger.warn("Caught exception while executing Gradle task named `properties`.", t)
+            false
+        }
+      val info = GradleProjectInfo(
+        buildEnv.getGradle.getGradleVersion,
+        project.getTasks.asScala.map(_.getName).toSeq,
+        hasAndroidPrefixGradleProperty
+      )
+      connection.close()
```
[the remark this answers is not in the document]
Comment (ursachec) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala:137]: :+1:

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala (hunk 2 above)
Comment (ursachec) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala]: :+1:

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 7]
```diff
@@ -102,114 +102,127 @@ object GradleDependencies {
     GradleInitScript(content, taskName)
   }
 
+  private def getGradleProjectInfo(projectDir: Path, projectName: String): Try[GradleProjectInfo] = {
```
Comment (ml86) [inline joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala:105]: For this function i would also switch to `Option` as return value.

Comment (ursachec) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala:105]: :+1:

**Review by mpollmeier — APPROVED** [review]

File: joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala [hunk 8]
```diff
@@ -102,114 +102,127 @@ object GradleDependencies {
...
+                          connection
+                            .newBuild()
+                            .forTasks(initScript.taskName)
+                            .withArguments("--init-script", initScriptFile.pathAsString)
+                            .setStandardOutput(stdoutStream)
+                            .setStandardError(stderrStream)
+                            .run()
+                        ) match {
+                          case Success(_) =>
+                            val result =
```
[the remark this answers is not in the document]
Comment (ursachec) [reply joern-cli/frontends/x2cpg/src/main/scala/io/joern/x2cpg/utils/dependency/GradleDependencies.scala]: yess

**Pull request merged**